File: apps/liga_app/lib/models/content.dart
import "package:flutter/material.dart";
import "package:google_fonts/google_fonts.dart";

class AppContent {
  final int version;
  final String disclaimer;
  final String homeScreenId;
  final List<AppScreen> screens;

  const AppContent({
    required this.version,
    required this.disclaimer,
    required this.homeScreenId,
    required this.screens,
  });

  factory AppContent.fromJson(Map<String, dynamic> json) {
    final screens = (json["screens"] as List? ?? [])
        .map((e) => AppScreen.fromJson(Map<String, dynamic>.from(e as Map)))
        .toList();
    return AppContent(
      version: (json["version"] as num?)?.toInt() ?? 1,
      disclaimer: json["disclaimer"] as String? ?? "",
      homeScreenId: json["homeScreenId"] as String? ?? (screens.isNotEmpty ? screens.first.id : ""),
      screens: screens,
    );
  }

  AppScreen? screenById(String id) {
    for (final s in screens) {
      if (s.id == id) return s;
    }
    return null;
  }
}

class AppScreen {
  final String id;
  final String title;
  final String? backgroundImage;
  /// Opacidade da imagem de fundo (0–1). Ausente / inválido = 1.
  final double backgroundOpacity;
  final List<ContentElement> elements;

  const AppScreen({
    required this.id,
    required this.title,
    this.backgroundImage,
    this.backgroundOpacity = 1,
    required this.elements,
  });

  factory AppScreen.fromJson(Map<String, dynamic> json) {
    final elements = (json["elements"] as List? ?? [])
        .map((e) => ContentElement.fromJson(Map<String, dynamic>.from(e as Map)))
        .toList();
    final bg = json["backgroundImage"] as String?;
    return AppScreen(
      id: json["id"] as String? ?? "",
      title: json["title"] as String? ?? "",
      backgroundImage: (bg != null && bg.isNotEmpty) ? bg : null,
      backgroundOpacity: normalizeOpacity(json["backgroundOpacity"]),
      elements: elements,
    );
  }
}

/// Aceita 0–1 ou 0–100 (legado). Default 1.
double normalizeOpacity(dynamic v) {
  if (v == null) return 1;
  if (v is! num) return 1;
  var n = v.toDouble();
  if (n > 1) n = n / 100;
  if (n.isNaN) return 1;
  return n.clamp(0.0, 1.0);
}

sealed class ContentElement {
  final String id;
  /// Layout opcional em % do canvas (0–100). Sem x/y = Column (auto).
  final double? x;
  final double? y;
  final double? w;
  /// Altura em % do canvas. Ausente = altura intrínseca.
  final double? h;
  /// Trava no editor; app ignora (só leitura).
  final bool locked;

  const ContentElement({
    required this.id,
    this.x,
    this.y,
    this.w,
    this.h,
    this.locked = false,
  });

  bool get hasLayout => x != null && y != null;

  factory ContentElement.fromJson(Map<String, dynamic> json) {
    final type = json["type"] as String? ?? "";
    switch (type) {
      case "button":
        return ButtonElement.fromJson(json);
      case "video":
        return VideoElement.fromJson(json);
      case "checklist":
        return ChecklistElement.fromJson(json);
      case "text":
        return TextElement.fromJson(json);
      case "image":
        return ImageElement.fromJson(json);
      default:
        return TextElement(id: json["id"] as String? ?? "unknown", content: "Elemento desconhecido: $type");
    }
  }

  static double? _pct(dynamic v) {
    if (v == null) return null;
    if (v is num) return v.toDouble();
    return null;
  }
}

class ButtonAction {
  final String type;
  final String target;
  const ButtonAction({required this.type, required this.target});
  factory ButtonAction.fromJson(Map<String, dynamic> json) => ButtonAction(
        type: json["type"] as String? ?? "navigate",
        target: json["target"] as String? ?? "",
      );
}

class ButtonElement extends ContentElement {
  final String label;
  final ButtonAction action;
  /// none | click | pop | beep | caminho de asset
  final String clickSound;
  final String bgColor;
  final String textColor;
  final String fontFamily;
  final double fontSize;
  final FontWeight fontWeight;
  final double borderRadius;
  final double opacity;
  final String? borderColor;
  final double borderWidth;
  final double paddingY;

  const ButtonElement({
    required super.id,
    required this.label,
    required this.action,
    this.clickSound = "none",
    this.bgColor = "#38bdf8",
    this.textColor = "#0f172a",
    this.fontFamily = "system",
    this.fontSize = 16,
    this.fontWeight = FontWeight.w600,
    this.borderRadius = 12,
    this.opacity = 1,
    this.borderColor,
    this.borderWidth = 0,
    this.paddingY = 12,
    super.x,
    super.y,
    super.w,
    super.h,
    super.locked,
  });

  factory ButtonElement.fromJson(Map<String, dynamic> json) => ButtonElement(
        id: json["id"] as String? ?? "",
        label: json["label"] as String? ?? "Botão",
        action: ButtonAction.fromJson(Map<String, dynamic>.from(json["action"] as Map? ?? {})),
        clickSound: (json["clickSound"] as String?)?.trim().isNotEmpty == true
            ? (json["clickSound"] as String).trim()
            : "none",
        bgColor: (json["bgColor"] as String?)?.trim().isNotEmpty == true
            ? (json["bgColor"] as String).trim()
            : "#38bdf8",
        textColor: (json["textColor"] as String?)?.trim().isNotEmpty == true
            ? (json["textColor"] as String).trim()
            : "#0f172a",
        fontFamily: json["fontFamily"] as String? ?? "system",
        fontSize: (json["fontSize"] as num?)?.toDouble() ?? 16,
        fontWeight: _parseWeight(json["fontWeight"], fallback: FontWeight.w600),
        borderRadius: (json["borderRadius"] as num?)?.toDouble() ?? 12,
        opacity: normalizeOpacity(json["opacity"] ?? 1),
        borderColor: (json["borderColor"] as String?)?.trim().isNotEmpty == true
            ? (json["borderColor"] as String).trim()
            : null,
        borderWidth: (json["borderWidth"] as num?)?.toDouble() ?? 0,
        paddingY: (json["paddingY"] as num?)?.toDouble() ?? 12,
        x: ContentElement._pct(json["x"]),
        y: ContentElement._pct(json["y"]),
        w: ContentElement._pct(json["w"]),
        h: ContentElement._pct(json["h"]),
        locked: json["locked"] == true,
      );

  Color get resolvedBg => TextElement._parseColor(bgColor) ?? const Color(0xFF38BDF8);
  Color get resolvedText => TextElement._parseColor(textColor) ?? const Color(0xFF0F172A);
  Color? get resolvedBorder =>
      borderColor != null ? TextElement._parseColor(borderColor) : null;

  TextStyle resolveLabelStyle() {
    final base = TextStyle(
      fontSize: fontSize,
      fontWeight: fontWeight,
      color: resolvedText,
    );
    switch (fontFamily) {
      case "Roboto":
        return GoogleFonts.roboto(textStyle: base);
      case "Open Sans":
        return GoogleFonts.openSans(textStyle: base);
      case "Lato":
        return GoogleFonts.lato(textStyle: base);
      case "Nunito":
        return GoogleFonts.nunito(textStyle: base);
      case "Montserrat":
        return GoogleFonts.montserrat(textStyle: base);
      case "system":
      default:
        return base;
    }
  }

  static FontWeight _parseWeight(dynamic v, {FontWeight fallback = FontWeight.w400}) {
    if (v == null) return fallback;
    final n = v is num ? v.toInt() : int.tryParse(v.toString()) ?? fallback.value;
    return FontWeight.values.firstWhere(
      (w) => w.value == n,
      orElse: () => fallback,
    );
  }
}

class VideoElement extends ContentElement {
  final String url;
  final String? title;
  const VideoElement({
    required super.id,
    required this.url,
    this.title,
    super.x,
    super.y,
    super.w,
    super.h,
    super.locked,
  });
  factory VideoElement.fromJson(Map<String, dynamic> json) => VideoElement(
        id: json["id"] as String? ?? "",
        url: json["url"] as String? ?? "",
        title: json["title"] as String?,
        x: ContentElement._pct(json["x"]),
        y: ContentElement._pct(json["y"]),
        w: ContentElement._pct(json["w"]),
        h: ContentElement._pct(json["h"]),
        locked: json["locked"] == true,
      );
}

class ChecklistItem {
  final String id;
  final String label;
  const ChecklistItem({required this.id, required this.label});
  factory ChecklistItem.fromJson(Map<String, dynamic> json) => ChecklistItem(
        id: json["id"] as String? ?? "",
        label: json["label"] as String? ?? "",
      );
}

class ChecklistElement extends ContentElement {
  final String title;
  final List<ChecklistItem> items;
  const ChecklistElement({
    required super.id,
    required this.title,
    required this.items,
    super.x,
    super.y,
    super.w,
    super.h,
    super.locked,
  });
  factory ChecklistElement.fromJson(Map<String, dynamic> json) => ChecklistElement(
        id: json["id"] as String? ?? "",
        title: json["title"] as String? ?? "Checklist",
        items: (json["items"] as List? ?? [])
            .map((e) => ChecklistItem.fromJson(Map<String, dynamic>.from(e as Map)))
            .toList(),
        x: ContentElement._pct(json["x"]),
        y: ContentElement._pct(json["y"]),
        w: ContentElement._pct(json["w"]),
        h: ContentElement._pct(json["h"]),
        locked: json["locked"] == true,
      );
}

class TextElement extends ContentElement {
  final String content;
  final String fontFamily;
  final double fontSize;
  final FontWeight fontWeight;
  final Color color;
  final TextAlign textAlign;
  final double lineHeight;

  const TextElement({
    required super.id,
    required this.content,
    this.fontFamily = "system",
    this.fontSize = 16,
    this.fontWeight = FontWeight.w400,
    this.color = const Color(0xFFE2E8F0),
    this.textAlign = TextAlign.left,
    this.lineHeight = 1.45,
    super.x,
    super.y,
    super.w,
    super.h,
    super.locked,
  });

  factory TextElement.fromJson(Map<String, dynamic> json) {
    return TextElement(
      id: json["id"] as String? ?? "",
      content: json["content"] as String? ?? "",
      fontFamily: json["fontFamily"] as String? ?? "system",
      fontSize: (json["fontSize"] as num?)?.toDouble() ?? 16,
      fontWeight: _parseWeight(json["fontWeight"]),
      color: _parseColor(json["color"]) ?? const Color(0xFFE2E8F0),
      textAlign: _parseAlign(json["textAlign"]),
      lineHeight: (json["lineHeight"] as num?)?.toDouble() ?? 1.45,
      x: ContentElement._pct(json["x"]),
      y: ContentElement._pct(json["y"]),
      w: ContentElement._pct(json["w"]),
      h: ContentElement._pct(json["h"]),
      locked: json["locked"] == true,
    );
  }

  TextStyle resolveStyle() {
    final base = TextStyle(
      fontSize: fontSize,
      fontWeight: fontWeight,
      color: color,
      height: lineHeight,
    );
    switch (fontFamily) {
      case "Roboto":
        return GoogleFonts.roboto(textStyle: base);
      case "Open Sans":
        return GoogleFonts.openSans(textStyle: base);
      case "Lato":
        return GoogleFonts.lato(textStyle: base);
      case "Nunito":
        return GoogleFonts.nunito(textStyle: base);
      case "Montserrat":
        return GoogleFonts.montserrat(textStyle: base);
      case "system":
      default:
        return base;
    }
  }

  static FontWeight _parseWeight(dynamic v) {
    if (v == null) return FontWeight.w400;
    final n = v is num ? v.toInt() : int.tryParse(v.toString()) ?? 400;
    return FontWeight.values.firstWhere(
      (w) => w.value == n,
      orElse: () => FontWeight.w400,
    );
  }

  static Color? _parseColor(dynamic v) {
    if (v is! String || v.isEmpty) return null;
    var s = v.trim();
    if (s.startsWith("#")) s = s.substring(1);
    if (s.length == 6) s = "FF$s";
    if (s.length != 8) return null;
    final n = int.tryParse(s, radix: 16);
    if (n == null) return null;
    return Color(n);
  }

  static TextAlign _parseAlign(dynamic v) {
    switch (v) {
      case "center":
        return TextAlign.center;
      case "right":
        return TextAlign.right;
      default:
        return TextAlign.left;
    }
  }
}

class ImageElement extends ContentElement {
  final String src;
  final String? alt;
  final String fit; // cover | contain
  final double? width;
  final double? height;
  final String? role; // logo | icon | photo

  const ImageElement({
    required super.id,
    required this.src,
    this.alt,
    this.fit = "contain",
    this.width,
    this.height,
    this.role,
    super.x,
    super.y,
    super.w,
    super.h,
    super.locked,
  });

  factory ImageElement.fromJson(Map<String, dynamic> json) => ImageElement(
        id: json["id"] as String? ?? "",
        src: json["src"] as String? ?? "",
        alt: json["alt"] as String?,
        fit: json["fit"] as String? ?? "contain",
        width: (json["width"] as num?)?.toDouble(),
        height: (json["height"] as num?)?.toDouble(),
        role: json["role"] as String?,
        x: ContentElement._pct(json["x"]),
        y: ContentElement._pct(json["y"]),
        w: ContentElement._pct(json["w"]),
        h: ContentElement._pct(json["h"]),
        locked: json["locked"] == true,
      );

  bool get isEmoji => src.startsWith("emoji:");
  String get emojiChar => isEmoji ? src.substring(6) : "";

  /// Caminho de asset Flutter para /media/arquivo → assets/media/arquivo
  String? get assetPath {
    if (src.startsWith("/media/")) {
      return "assets/media/${src.substring("/media/".length)}";
    }
    if (src.startsWith("media/")) {
      return "assets/$src";
    }
    return null;
  }

  bool get isNetwork =>
      src.startsWith("http://") || src.startsWith("https://");
}

/// Utilitário para plano de fundo da tela (mesmas regras de ImageElement.src).
class MediaPath {
  static bool isEmoji(String src) => src.startsWith("emoji:");
  static String emojiChar(String src) =>
      isEmoji(src) ? src.substring(6) : "";

  static String? assetPath(String src) {
    if (src.startsWith("/media/")) {
      return "assets/media/${src.substring("/media/".length)}";
    }
    if (src.startsWith("media/")) {
      return "assets/$src";
    }
    return null;
  }

  static bool isNetwork(String src) =>
      src.startsWith("http://") || src.startsWith("https://");
}

/// Resolve preset de som → asset path (ou null se none).
String? clickSoundAsset(String? clickSound) {
  if (clickSound == null || clickSound.isEmpty || clickSound == "none") return null;
  const presets = {
    "click": "assets/sounds/click.wav",
    "pop": "assets/sounds/pop.wav",
    "beep": "assets/sounds/beep.wav",
  };
  if (presets.containsKey(clickSound)) return presets[clickSound];
  if (clickSound.startsWith("assets/")) return clickSound;
  return null;
}

File: apps/editor/src/types/content.ts
export type NavigateAction = { type: "navigate"; target: string };
export type OpenUrlAction = { type: "openUrl"; target: string };
export type ButtonAction = NavigateAction | OpenUrlAction;

/** Layout opcional em % do canvas do celular (0–100). Sem x/y = empilhar (auto). */
export type ElementLayout = {
  x?: number;
  y?: number;
  w?: number;
  /** Altura em % do canvas. Ausente = altura do conteúdo. */
  h?: number;
  /** Trava posição/tamanho no canvas (arrasto/resize). */
  locked?: boolean;
};

/** Presets de som do botão (assets no Flutter). */
export type ClickSoundId = "none" | "click" | "pop" | "beep";

export const CLICK_SOUND_OPTIONS: { id: ClickSoundId; label: string }[] = [
  { id: "none", label: "Nenhum" },
  { id: "click", label: "Clique" },
  { id: "pop", label: "Pop" },
  { id: "beep", label: "Beep" },
];

/** Forma do botão → borderRadius em px (pill ≈ 999). */
export type ButtonShapeId = "retangulo" | "arredondado" | "pill";

export const BUTTON_SHAPE_OPTIONS: { id: ButtonShapeId; label: string; radius: number }[] = [
  { id: "retangulo", label: "Retângulo", radius: 4 },
  { id: "arredondado", label: "Arredondado", radius: 12 },
  { id: "pill", label: "Pill", radius: 999 },
];

export const BUTTON_STYLE_DEFAULTS = {
  bgColor: "#38bdf8",
  textColor: "#0f172a",
  fontFamily: "system" as const,
  fontSize: 16,
  fontWeight: 600,
  borderRadius: 12,
  opacity: 1,
  borderColor: undefined as string | undefined,
  borderWidth: 0,
  paddingY: 12,
};

export type ButtonElement = {
  id: string;
  type: "button";
  label: string;
  action: ButtonAction;
  /** Som ao tocar no app publicado (none/click/pop/beep). Preferências do usuário ficam no app. */
  clickSound?: ClickSoundId | string;
  /** Cor de fundo (hex). Default #38bdf8. */
  bgColor?: string;
  /** Cor do texto (hex). */
  textColor?: string;
  fontFamily?: FontFamilyId | string;
  fontSize?: number;
  fontWeight?: number;
  /** Raio em px (4 = retângulo, 12 = arredondado, 999 = pill). */
  borderRadius?: number;
  /** Opacidade 0–1. */
  opacity?: number;
  borderColor?: string;
  borderWidth?: number;
  /** Padding vertical em px. */
  paddingY?: number;
} & ElementLayout;

export function buttonShapeFromRadius(r?: number): ButtonShapeId {
  if (r == null) return "arredondado";
  if (r >= 100) return "pill";
  if (r <= 6) return "retangulo";
  return "arredondado";
}

export function radiusForShape(shape: ButtonShapeId): number {
  return BUTTON_SHAPE_OPTIONS.find((o) => o.id === shape)?.radius ?? 12;
}

/** Resolve estilos do botão com defaults (rascunhos antigos). */
export function resolveButtonStyle(el: ButtonElement) {
  const d = BUTTON_STYLE_DEFAULTS;
  return {
    bgColor: el.bgColor || d.bgColor,
    textColor: el.textColor || d.textColor,
    fontFamily: el.fontFamily || d.fontFamily,
    fontSize: el.fontSize ?? d.fontSize,
    fontWeight: el.fontWeight ?? d.fontWeight,
    borderRadius: el.borderRadius ?? d.borderRadius,
    opacity: normalizeOpacity(el.opacity ?? d.opacity),
    borderColor: el.borderColor,
    borderWidth: el.borderWidth ?? d.borderWidth,
    paddingY: el.paddingY ?? d.paddingY,
  };
}

/** Templates clínicos de 1 toque (Primário / Alerta / Secundário). */
export type ButtonTemplateId = "primario" | "alerta" | "secundario";

export type ButtonStylePatch = {
  bgColor: string;
  textColor: string;
  borderRadius: number;
  borderWidth: number;
  borderColor?: string;
  opacity: number;
  fontWeight: number;
  fontSize: number;
  paddingY: number;
};

export const BUTTON_TEMPLATES: {
  id: ButtonTemplateId;
  label: string;
  hint: string;
  patch: ButtonStylePatch;
}[] = [
  {
    id: "primario",
    label: "Primário",
    hint: "Ação principal — azul/verde clínico",
    patch: {
      bgColor: "#0d9488",
      textColor: "#ffffff",
      borderRadius: 12,
      borderWidth: 0,
      opacity: 1,
      fontWeight: 600,
      fontSize: 16,
      paddingY: 12,
    },
  },
  {
    id: "alerta",
    label: "Alerta",
    hint: "Urgência — vermelho/âmbar",
    patch: {
      bgColor: "#dc2626",
      textColor: "#ffffff",
      borderRadius: 12,
      borderWidth: 2,
      borderColor: "#f59e0b",
      opacity: 1,
      fontWeight: 700,
      fontSize: 16,
      paddingY: 12,
    },
  },
  {
    id: "secundario",
    label: "Secundário",
    hint: "Outline — secundário",
    patch: {
      bgColor: "#0f172a",
      textColor: "#7dd3fc",
      borderRadius: 12,
      borderWidth: 2,
      borderColor: "#38bdf8",
      opacity: 1,
      fontWeight: 600,
      fontSize: 16,
      paddingY: 12,
    },
  },
];

function hexToRgb(hex: string): [number, number, number] | null {
  let s = hex.trim();
  if (s.startsWith("#")) s = s.slice(1);
  if (s.length === 3) s = s.split("").map((c) => c + c).join("");
  if (s.length !== 6) return null;
  const n = Number.parseInt(s, 16);
  if (Number.isNaN(n)) return null;
  return [(n >> 16) & 255, (n >> 8) & 255, n & 255];
}

function relativeLuminance(hex: string): number {
  const rgb = hexToRgb(hex);
  if (!rgb) return 0.5;
  const lin = (c: number) => {
    const x = c / 255;
    return x <= 0.03928 ? x / 12.92 : Math.pow((x + 0.055) / 1.055, 2.4);
  };
  const [r, g, b] = rgb.map(lin);
  return 0.2126 * r + 0.7152 * g + 0.0722 * b;
}

/** Razão de contraste WCAG (bg vs texto). */
export function buttonContrastRatio(bg: string, fg: string): number {
  const L1 = relativeLuminance(bg);
  const L2 = relativeLuminance(fg);
  const light = Math.max(L1, L2);
  const dark = Math.min(L1, L2);
  return (light + 0.05) / (dark + 0.05);
}

/** true se contraste < 4.5:1 (aviso amigável no inspetor). */
export function buttonContrastWarn(bg: string, fg: string): boolean {
  return buttonContrastRatio(bg, fg) < 4.5;
}

export type VideoElement = {
  id: string;
  type: "video";
  url: string;
  title?: string;
} & ElementLayout;

export type ChecklistItem = { id: string; label: string };

export type ChecklistElement = {
  id: string;
  type: "checklist";
  title: string;
  items: ChecklistItem[];
} & ElementLayout;

/** Fontes livres (system + Google Fonts). Sem fontes licenciadas Microsoft. */
export type FontFamilyId = "system" | "Roboto" | "Open Sans" | "Lato" | "Nunito" | "Montserrat";

export const FONT_FAMILY_OPTIONS: { id: FontFamilyId; label: string; css: string }[] = [
  { id: "system", label: "Sistema", css: "system-ui, -apple-system, Segoe UI, Roboto, sans-serif" },
  { id: "Roboto", label: "Roboto", css: "Roboto, sans-serif" },
  { id: "Open Sans", label: "Open Sans", css: "'Open Sans', sans-serif" },
  { id: "Lato", label: "Lato", css: "Lato, sans-serif" },
  { id: "Nunito", label: "Nunito", css: "Nunito, sans-serif" },
  { id: "Montserrat", label: "Montserrat", css: "Montserrat, sans-serif" },
];

export type TextAlignId = "left" | "center" | "right";

export type TextElement = {
  id: string;
  type: "text";
  content: string;
  fontFamily?: FontFamilyId | string;
  fontSize?: number;
  fontWeight?: number;
  color?: string;
  textAlign?: TextAlignId;
  lineHeight?: number;
} & ElementLayout;

export type ImageFit = "cover" | "contain";
export type ImageRole = "logo" | "icon" | "photo";

export type ImageElement = {
  id: string;
  type: "image";
  src: string;
  alt?: string;
  fit?: ImageFit;
  width?: number;
  height?: number;
  role?: ImageRole;
} & ElementLayout;

export type ContentElement =
  | ButtonElement
  | VideoElement
  | ChecklistElement
  | TextElement
  | ImageElement;

export type Screen = {
  id: string;
  title: string;
  /** URL http(s) ou caminho /media/... */
  backgroundImage?: string;
  /**
   * Opacidade da imagem de fundo (0–1). Ausente = 1.
   * No editor o slider mostra 0–100%.
   */
  backgroundOpacity?: number;
  elements: ContentElement[];
};

export type AppContent = {
  version: number;
  disclaimer: string;
  homeScreenId: string;
  screens: Screen[];
};

/** Atalhos de ícone (emoji) — src no formato emoji:XXX */
export const ICON_PRESETS: { label: string; emoji: string }[] = [
  { label: "Ambulância", emoji: "🚑" },
  { label: "Hospital", emoji: "🏥" },
  { label: "Coração", emoji: "❤️" },
  { label: "Estetoscópio", emoji: "🩺" },
  { label: "Medicina", emoji: "⚕️" },
  { label: "Alerta", emoji: "⚠️" },
  { label: "Checklist", emoji: "📋" },
  { label: "OK", emoji: "✅" },
  { label: "Estrela", emoji: "⭐" },
  { label: "Telefone", emoji: "📱" },
];

export const TOOL_TYPES: { type: ContentElement["type"]; label: string; icon: string }[] = [
  { type: "button", label: "Botão", icon: "🔘" },
  { type: "image", label: "Imagem", icon: "🖼️" },
  { type: "video", label: "Vídeo", icon: "▶️" },
  { type: "checklist", label: "Checklist", icon: "☑️" },
  { type: "text", label: "Texto", icon: "🔤" },
];

export function emojiSrc(emoji: string): string {
  return "emoji:" + emoji;
}

export function isEmojiSrc(src: string): boolean {
  return src.startsWith("emoji:");
}

export function emojiFromSrc(src: string): string {
  return src.startsWith("emoji:") ? src.slice(6) : "";
}

/** Resolve caminho de mídia para uso na UI do editor (preview). */
export function resolveMediaSrc(src: string): string {
  if (!src) return "";
  if (isEmojiSrc(src)) return src;
  if (src.startsWith("http://") || src.startsWith("https://") || src.startsWith("data:")) {
    return src;
  }
  if (src.startsWith("/media/")) return src;
  if (src.startsWith("media/")) return "/" + src;
  return src;
}

export function hasLayout(el: ElementLayout): boolean {
  return typeof el.x === "number" && typeof el.y === "number";
}

export function defaultWidthFor(type: ContentElement["type"]): number {
  if (type === "image") return 50;
  if (type === "text") return 70;
  return 80;
}

/** Altura padrão opcional (%). Imagem ganha h; outros ficam auto até o usuário definir. */
export function defaultHeightFor(type: ContentElement["type"]): number | undefined {
  if (type === "image") return 25;
  if (type === "button") return 8;
  return undefined;
}

/** Normaliza opacidade 0–1 (aceita legado 0–100). Default 1. */
export function normalizeOpacity(v: number | undefined | null): number {
  if (v == null || Number.isNaN(Number(v))) return 1;
  let n = Number(v);
  if (n > 1) n = n / 100;
  return Math.max(0, Math.min(1, n));
}

export function fontCss(family?: string): string {
  const found = FONT_FAMILY_OPTIONS.find((f) => f.id === family);
  return found?.css ?? FONT_FAMILY_OPTIONS[0].css;
}

export function createElementOfType(
  type: ContentElement["type"],
  layout?: ElementLayout
): ContentElement {
  const pos: ElementLayout = {
    x: layout?.x,
    y: layout?.y,
    w: layout?.w ?? defaultWidthFor(type),
    h: layout?.h ?? defaultHeightFor(type),
    locked: layout?.locked,
  };
  if (type === "button") {
    const d = BUTTON_STYLE_DEFAULTS;
    return {
      id: uid("btn"),
      type,
      label: "Botão",
      action: { type: "navigate", target: "" },
      clickSound: "none",
      bgColor: d.bgColor,
      textColor: d.textColor,
      fontFamily: d.fontFamily,
      fontSize: d.fontSize,
      fontWeight: d.fontWeight,
      borderRadius: d.borderRadius,
      opacity: d.opacity,
      borderWidth: d.borderWidth,
      paddingY: d.paddingY,
      ...pos,
    };
  }
  if (type === "video") {
    return { id: uid("vid"), type, url: "", title: "Vídeo", ...pos };
  }
  if (type === "checklist") {
    return { id: uid("chk"), type, title: "Checklist", items: [{ id: uid("item"), label: "Item" }], ...pos };
  }
  if (type === "image") {
    return {
      id: uid("img"),
      type: "image",
      src: "",
      alt: "",
      fit: "contain",
      role: "photo",
      ...pos,
    };
  }
  return {
    id: uid("txt"),
    type: "text",
    content: "Texto",
    fontFamily: "system",
    fontSize: 16,
    fontWeight: 400,
    color: "#e2e8f0",
    textAlign: "left",
    lineHeight: 1.45,
    ...pos,
  };
}

export function createEmptyContent(): AppContent {
  return {
    version: 1,
    disclaimer: "Conteúdo educacional da LAUEM. Não substitui protocolos clínicos oficiais.",
    homeScreenId: "home",
    screens: [{ id: "home", title: "Início", elements: [] }],
  };
}

export function uid(prefix = "id"): string {
  return prefix + "_" + Math.random().toString(36).slice(2, 9);
}

/** Clamp layout % */
export function clampPct(n: number, min = 0, max = 100): number {
  return Math.max(min, Math.min(max, Math.round(n * 10) / 10));
}

/** Duplica elemento com novo id (itens de checklist também). */
export function duplicateElement(el: ContentElement): ContentElement {
  const base = { ...el, id: uid(el.type.slice(0, 3)) };
  if (base.type === "checklist") {
    return {
      ...base,
      items: base.items.map((it) => ({ ...it, id: uid("item") })),
    };
  }
  // offset leve pra não ficar em cima
  if (typeof base.x === "number") base.x = clampPct(base.x + 3, 0, 92);
  if (typeof base.y === "number") base.y = clampPct(base.y + 3, 0, 92);
  return base;
}
